libcarina_test.go

```diff
@@ -0,0 +1,223 @@
+package libcarina
+
+import (
+	"net/http"
+	"reflect"
+	"testing"
+
+	"fmt"
+	"github.com/pkg/errors"
+	"net/http/httptest"
+)
+
+const mockUsername = "test-user"
+const mockAPIKey = "1234"
+const mockRegion = "DFW"
+const mockToken = ""
+
+const microversionUnsupportedJSON = `{"errors":[{"code":"make-coe-api.microverion-unsupported","detail":"If the api-version header is sent, it must be in the format 'rax:container X.Y' where 1.0 <= X.Y <= 1.0","links":[{"href":"https://getcarina.com/docs/","rel":"help"}],"max_version":"1.0","min_version":"1.0","request_id":"620c8d81-b8f9-4bb0-952b-6d08ae42eda0","status":406,"title":"Microversion unsupported"}]}`
+
+type handler func(w http.ResponseWriter, r *http.Request)
+
+func identityHandler(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+
+	switch r.RequestURI {
+	case "/v2.0/tokens":
+		fmt.Fprintln(w, `{"access":{"serviceCatalog":[{"endpoints":[{"tenantId":"963451","publicURL":"https:\/\/api.dfw.getcarina.com","region":"DFW"}],"name":"cloudContainer","type":"rax:container"}],"user":{"name":"fake-user","id":"fake-userid"},"token":{"expires":"3000-01-01T12:00:00Z","id":"fake-token","tenant":{"name":"fake-tenantname","id":"fake-tenantid"}}}}`)
+	default:
+		w.WriteHeader(404)
+		fmt.Fprintln(w, "unexpected request: "+r.RequestURI)
+	}
+}
+
+func microversionUnsupportedHandler(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(406)
+	fmt.Fprintln(w, microversionUnsupportedJSON)
+}
+
+func createMockCarina(h handler) (*httptest.Server, *httptest.Server) {
+	return httptest.NewServer(http.HandlerFunc(h)), httptest.NewServer(http.HandlerFunc(identityHandler))
+}
+
+func createMockCarinaClient(identity string, endpoint string) (*CarinaClient, error) {
+	client, err := NewClient(mockUsername, mockAPIKey, mockRegion, identity, mockToken, endpoint)
+	if err != nil {
+		return client, err
+	}
+	client.Endpoint = endpoint
+	return client, nil
+}
+
+func assertMicroversionUnsupportedHandled(t *testing.T, err error) {
+	if err == nil {
+		t.Error("expected to get error")
+	}
+	cause := errors.Cause(err)
+	if httpErr, ok := cause.(HTTPErr); ok {
+		if httpErr.StatusCode != 406 {
+			t.Error("expected StatusCode of 406, got", httpErr.StatusCode)
+		}
+	} else {
+		t.Error("expected to get HTTPErr, got", reflect.TypeOf(err))
+	}
+}
+
+func TestMicroversionUnsupportedNewRequest(t *testing.T) {
```

Comment:
I think it's some extra assurance about all things calling the Carina API. Sort of asserts that the expectations of `NewRequest` for other functions that're using it. That way you can tell, if you start seeing failures, whether it's in `NewRequest`, or something calling it.